stl/inc/ranges

```diff
@@ -2490,6 +2490,428 @@ namespace ranges {
+                conditional_t<derived_from<_BaseCategory, forward_iterator_tag>, forward_iterator_tag, _BaseCategory>;
+        };
+
+        template <bool _Const>
+        class _Inner_iter : public _Inner_iter_base<_Maybe_const<_Const, _Vw>> {
+        private:
+            using _BaseTy = _Maybe_const<_Const, _Vw>;
+
+            _Outer_iter<_Const> _It{};
+            bool _Incremented = false;
+
+            _NODISCARD constexpr bool _Equal(const _Inner_iter& _Right) const {
+                return _It._Get_current() == _Right._It._Get_current();
+            }
+
+            _NODISCARD constexpr iterator_t<_BaseTy>& _Get_current() noexcept {
+                return _It._Get_current();
+            }
+
+            _NODISCARD constexpr const iterator_t<_BaseTy>& _Get_current() const noexcept {
+                return _It._Get_current();
+            }
+
+            _NODISCARD constexpr bool _At_end() const {
+                auto _Pat_pos       = _RANGES begin(_It._Parent->_Pattern);
+                const auto _Pat_end = _RANGES end(_It._Parent->_Pattern);
+                auto _Last          = _RANGES end(_It._Parent->_Range);
+                if constexpr (_Tiny_range<_Pat>) {
+                    const auto& _Cur = _It._Get_current();
+                    if (_Cur == _Last) {
+                        return true;
+                    }
+
+                    if (_Pat_pos == _Pat_end) {
+                        return _Incremented;
+                    }
+                    return *_Cur == *_Pat_pos;
+                } else {
+                    auto _Cur = _It._Get_current();
+                    if (_Cur == _Last) {
+                        return true;
+                    }
+
+                    if (_Pat_pos == _Pat_end) {
+                        return _Incremented;
+                    }
+
+                    do {
+                        if (*_Cur != *_Pat_pos) {
+                            return false;
+                        }
+
+                        if (++_Pat_pos == _Pat_end) {
+                            return true;
+                        }
+                    } while (++_Cur != _Last);
+                    return false;
+                }
+            }
+
+
+        public:
+            using iterator_concept = typename _Outer_iter<_Const>::iterator_concept;
+            using value_type       = range_value_t<_BaseTy>;
+            using difference_type  = range_difference_t<_BaseTy>;
+
+            _Inner_iter() = default;
+            constexpr explicit _Inner_iter(_Outer_iter<_Const> _It_) noexcept(
+                is_nothrow_move_constructible_v<_Outer_iter<_Const>>) // strengthened
+                : _It{_STD move(_It_)} {}
+
+            _NODISCARD constexpr decltype(auto) operator*() const {
+                return *_It._Get_current();
+            }
+
+            constexpr _Inner_iter& operator++() {
+                _Incremented = true;
+                if constexpr (!forward_range<_BaseTy>) {

```

Comment:
> Should this be only one `if constexpr`?

Nope: `_Pat::size()` might be a non-constant expression if `forward_range<_BaseTy>`. (!forward_range<_BaseTy>` here really means `_Tiny_range<_Pat>` - as required by the constraints on `split_view`. Don't ask me we the Standard wording says `!forward_range` instead of `_Tiny_range` which seems much clearer; I'll change this.

Reply:
The behavior is different if the range is forward and the pattern is tiny (and empty), and I believe using _`tiny-range`_ here damages the semantics of `operator==(const inner-iterator&, const inner-iterator&)`.

Reply:
> The behavior is different if the range is forward and the pattern is tiny (and empty), and I believe using _`tiny-range`_ here damages the semantics of `operator==(const inner-iterator&, const inner-iterator&)`.

Agreed - I'm going to stop making changes without adding specific test coverage.